pkg/controller/machinedeployment/util/util.go

```diff
@@ -0,0 +1,691 @@
+func getMachineSetFraction(ms v1alpha1.MachineSet, d v1alpha1.MachineDeployment) int32 {
+	// If we are scaling down to zero then the fraction of this machine set is its whole size (negative)
+	if *(d.Spec.Replicas) == int32(0) {
+		return -*(ms.Spec.Replicas)
+	}
+
+	deploymentReplicas := *(d.Spec.Replicas) + MaxSurge(d)
+	annotatedReplicas, ok := getMaxReplicasAnnotation(&ms)
+	if !ok {
+		// If we cannot find the annotation then fallback to the current deployment size. Note that this
+		// will not be an accurate proportion estimation in case other machine sets have different values
+		// which means that the deployment was scaled at some point but we at least will stay in limits
+		// due to the min-max comparisons in getProportion.
+		annotatedReplicas = d.Status.Replicas
+	}
+
+	// We should never proportionally scale up from zero which means ms.spec.replicas and annotatedReplicas
+	// will never be zero here.
+	newMSsize := (float64(*(ms.Spec.Replicas) * deploymentReplicas)) / float64(annotatedReplicas)
+	return integer.RoundToInt32(newMSsize) - *(ms.Spec.Replicas)
+}
+
+// EqualIgnoreHash returns true if two given machineTemplateSpec are equal, ignoring the diff in value of Labels[machine-template-hash]
+// We ignore machine-template-hash because:
+// 1. The hash result would be different upon machineTemplateSpec API changes
+//    (e.g. the addition of a new field will cause the hash code to change)
+// 2. The deployment template won't have hash labels
+func EqualIgnoreHash(template1, template2 *v1alpha1.MachineTemplateSpec) bool {
+	t1Copy := template1.DeepCopy()
+	t2Copy := template2.DeepCopy()
+	// Remove hash labels from template.Labels before comparing
+	delete(t1Copy.Labels, v1alpha1.DefaultMachineDeploymentUniqueLabelKey)
+	delete(t2Copy.Labels, v1alpha1.DefaultMachineDeploymentUniqueLabelKey)
+	return apiequality.Semantic.DeepEqual(t1Copy, t2Copy)
+}
+
+// FindNewMachineSet returns the new MS this given deployment targets (the one with the same machine template).
+func FindNewMachineSet(deployment *v1alpha1.MachineDeployment, msList []*v1alpha1.MachineSet) *v1alpha1.MachineSet {
+	sort.Sort(MachineSetsByCreationTimestamp(msList))
+	for i := range msList {
+		if EqualIgnoreHash(&msList[i].Spec.Template, &deployment.Spec.Template) {
+			// In rare cases, such as after cluster upgrades, Deployment may end up with
+			// having more than one new MachineSets that have the same template as its template,
+			// see https://github.com/kubernetes/kubernetes/issues/40415
+			// We deterministically choose the oldest new MachineSet.
+			return msList[i]
+		}
+	}
+	// new MachineSet does not exist.
+	return nil
+}
+
+// FindOldMachineSets returns the old machine sets targeted by the given Deployment, with the given slice of MSes.
+// Note that the first set of old machine sets doesn't include the ones with no machines, and the second set of old machine sets include all old machine sets.
```

Comment:
Suggest rewording that this returns a list of all old machine sets with replicas scaled to 0, and a list includes all old machine sets regardless of replicas.

Reply:
done.